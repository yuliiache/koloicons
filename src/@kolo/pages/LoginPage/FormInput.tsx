import Input from '@kolo/uiKit/Input/Input';
import {InputType} from '@kolo/uiKit/Input/InputType';
import {FC} from 'react';
import {Field} from 'react-final-form';

interface Props {
  label?: string;
  placeholder?: string;
  isDisabled?: boolean;
  hint?: string;
  hasAutoFocus?: boolean;
  id: string;
  name: string;
  inputType: InputType;
}

const FormInput: FC<Props> = ({name, inputType, id, hasAutoFocus = false, label, isDisabled = false}) => {
  return (
    <Field
      name={name}
      render={({input: {value, onChange}, meta: {error, touched}}) => (
        <Input
          id={id}
          onChange={onChange}
          value={value}
          type={inputType}
          hasAutoFocus={hasAutoFocus}
          label={label}
          errorMessage={touched && error}
          isDisabled={isDisabled}
        />
      )}
    />
  );
};

export default FormInput;
