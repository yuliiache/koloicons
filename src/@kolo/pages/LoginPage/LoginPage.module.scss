@import "@kolo/design";

.container {
  box-sizing: border-box;
  width: 100%;
  max-width: 1024px;
  margin: 0 auto;
  padding: 0 16px;

  &Xs {
    max-width: 480px;
  }
}

.authorization {
  padding: 24px;
  background-color: $LightBG;
  border-radius: 4px;
  box-shadow: $primaryBoxShadowAuth, $secondaryBoxShadowAuth;

  &Form {
    margin-bottom: 24px;
  }
  &Footer {
    margin-top: 0;
    margin-bottom: 10px;
    text-align: center;
    font-size: 0.6rem;
    line-height: 1.33;
    color: $GreyFontBtn;
  }
}

.headerName {
  text-align: center;
  color: $secFontColor;
  margin-top: 0;
  margin-bottom: 26px;
  font-size: 0.8rem;
}
