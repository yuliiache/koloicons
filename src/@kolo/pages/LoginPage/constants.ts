export const LOGIN_USER_KEY = 'loginUser';
