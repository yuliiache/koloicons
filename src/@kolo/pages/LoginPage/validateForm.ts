import {validateEmail, validatePassword} from '@kolo/services/validators/commonValidators';

import type {Value} from './LoginPage';

interface Error {
  email?: string;
  password?: string;
}

export const initalValues: Value = {
  email: '',
  password: '',
};

export const validateForm = (value: Value): Error => {
  const {email, password} = value;

  const errors: Error = {};

  errors.email = validateEmail(email);
  errors.password = validatePassword(password);

  return errors;
};
