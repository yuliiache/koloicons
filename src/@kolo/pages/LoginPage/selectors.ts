import {RootState} from 'app/store';
import {Selector} from 'react-redux';
import {createSelector} from 'reselect';

import {LOGIN_USER_KEY} from './constants';
import {LoginState, UserData} from './slices/loginUserSlice';
import {ErrorCode} from './types/ErrorCode';

const getUser: Selector<RootState, LoginState> = (state) => state[LOGIN_USER_KEY];

export const userSelector: Selector<RootState, UserData> = createSelector(
  getUser,
  (loginUser: LoginState): UserData => loginUser.user || {name: '', lastName: '', email: '', id: '', isPro: false}
);

export const userIdSelector: Selector<RootState, string | undefined> = createSelector(
  getUser,
  (loginUser: LoginState): string | undefined => loginUser.user?.id
);

export const isAuthenticatedSelector: Selector<RootState, boolean> = createSelector(
  getUser,
  (loginUser: LoginState): boolean => loginUser.isAuthenticated
);

export const errorCodeSelector: Selector<RootState, ErrorCode> = createSelector(
  getUser,
  (loginUser: LoginState): ErrorCode => loginUser.errorCode
);
