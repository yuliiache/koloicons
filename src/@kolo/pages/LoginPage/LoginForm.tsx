import Button from '@kolo/uiKit/Button/Button';
import {ButtonSize, ButtonType} from '@kolo/uiKit/Button/constants';
import {InputType} from '@kolo/uiKit/Input/InputType';
import {FC} from 'react';
import {Form} from 'react-final-form';

import FormInput from './FormInput';
import type {Value} from './LoginPage';
import style from './LoginPage.module.scss';
import {initalValues, validateForm} from './validateForm';

const {authorizationForm} = style;

interface OnSubmit {
  onSubmit: (value: Value) => void;
}

const LoginForm: FC<OnSubmit> = ({onSubmit}) => {
  return (
    <Form
      initialValues={initalValues}
      validate={validateForm}
      onSubmit={onSubmit}
      render={({handleSubmit}) => (
        <form
          onSubmit={handleSubmit}
          className={authorizationForm}
        >
          <FormInput
            name={'email'}
            inputType={InputType.EMAIL}
            id={'email'}
            hasAutoFocus
            label={'Email'}
          />
          <FormInput
            name={'password'}
            inputType={InputType.PASSWORD}
            id={'password'}
            label={'Password'}
          />
          <Button
            text="Sign In"
            type={ButtonType.SUBMIT}
            size={ButtonSize.FULL_WIDTH}
          />
        </form>
      )}
    />
  );
};

export default LoginForm;
