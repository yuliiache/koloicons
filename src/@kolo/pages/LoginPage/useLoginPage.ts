import {useAppDispatch, useAppSelector} from 'app/hooks';
import {AppRoute} from 'constants/AppRoute';
import {useEffect} from 'react';
import {useNavigate} from 'react-router';

import {errorCodeSelector, isAuthenticatedSelector} from './selectors';
import {loginUserStartAction} from './slices/loginUserSlice';
import {ErrorCode} from './types/ErrorCode';
import {LoginFormValue} from './types/LoginFormValue';

type HandleSubmit = (value: LoginFormValue) => void;

interface LoginPageHookReturn {
  isAuthenticationError: boolean;
  handleSubmit: HandleSubmit;
}

const useLoginPage = (): LoginPageHookReturn => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const isAuthenticated = useAppSelector(isAuthenticatedSelector);
  const errorCode: ErrorCode = useAppSelector(errorCodeSelector);
  const isAuthenticationError = errorCode === 400;

  useEffect(() => {
    if (isAuthenticated) {
      navigate(AppRoute.ICONS);
    }
  }, [isAuthenticated]);

  const handleSubmit: HandleSubmit = (value) => {
    dispatch(loginUserStartAction(value));
  };

  return {isAuthenticationError, handleSubmit};
};

export default useLoginPage;
