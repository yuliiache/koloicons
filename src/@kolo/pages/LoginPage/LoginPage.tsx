import {ButtonColor, ButtonSize} from '@kolo/uiKit/Button/constants';
import ButtonLink from '@kolo/uiKit/ButtonLink/ButtonLink';
import classnames from 'classnames';
import {AppRoute} from 'constants/AppRoute';
import {FC} from 'react';

import LoginForm from './LoginForm';
import style from './LoginPage.module.scss';

const {authorization, headerName, container, containerXs, authorizationFooter} = style;

export interface Value {
  email: string;
  password: string;
}

const LoginPage: FC = () => {
  const handleSubmit = (value: Value) => {
    console.log(value);
  };

  return (
    <div className={classnames(container, containerXs)}>
      <section className={authorization}>
        <h2 className={headerName}>Login</h2>
        <LoginForm onSubmit={handleSubmit} />
        <p className={authorizationFooter}>Create an account to use KOLO Icons for free!</p>
        <ButtonLink
          url={AppRoute.REGISTER}
          text="Sign Up"
          color={ButtonColor.BASIC}
          size={ButtonSize.FULL_WIDTH}
        />
      </section>
    </div>
  );
};

export default LoginPage;
