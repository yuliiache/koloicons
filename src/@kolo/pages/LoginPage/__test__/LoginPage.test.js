import {BrowserRouter} from 'react-router-dom';
import LoginPage from '@kolo/pages/LoginPage/LoginPage';
import {fireEvent, render, screen} from '@testing-library/react';
import {AppRoute} from 'constants/AppRoute';

describe('LoginPage', () => {
  it('should create snapshot successfully', () => {
    const {asFragment} = render(<LoginPage />, {wrapper: BrowserRouter});

    expect(asFragment()).toMatchSnapshot();
  });

  it('should navigate to register page', () => {
    render(<LoginPage />, {wrapper: BrowserRouter});

    const button = screen.getByRole('button', {name: /Sign Up/i});
    fireEvent.click(button);

    expect(window.location.pathname).toBe(AppRoute.REGISTER);
  });
});
