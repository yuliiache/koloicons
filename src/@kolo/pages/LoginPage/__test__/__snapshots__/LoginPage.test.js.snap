// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should create snapshot successfully 1`] = `
<DocumentFragment>
  <div
    class="container containerXs"
  >
    <section
      class="authorization"
    >
      <h2
        class="headerName"
      >
        Login
      </h2>
      <form
        class="authorizationForm"
      >
        <div
          class="inputGroup"
        >
          <label
            class="inputGroupLabel"
            for="email"
          >
            Email
          </label>
          <input
            class="inputGroupInput"
            id="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="inputGroup"
        >
          <label
            class="inputGroupLabel"
            for="password"
          >
            Password
          </label>
          <input
            class="inputGroupInput"
            id="password"
            type="password"
            value=""
          />
        </div>
        <button
          class="button button-full-width button-primary"
          data-testid="buttonComponent"
          type="submit"
        >
          <span
            data-testid="buttonComponentText"
          >
            Sign In
          </span>
        </button>
      </form>
      <p
        class="authorizationFooter"
      >
        Create an account to use KOLO Icons for free!
      </p>
      <a
        data-testid="buttonLinkComponent"
        href="/register"
      >
        <button
          class="button button-full-width button-basic"
          data-testid="buttonComponent"
          type="button"
        >
          <span
            data-testid="buttonComponentText"
          >
            Sign Up
          </span>
        </button>
      </a>
    </section>
  </div>
</DocumentFragment>
`;
